modules/core/shared/src/main/scala/coursier/core/Version.scala

```diff
@@ -65,180 +68,108 @@ object Version {
-          case _ => (None, s)
-        }
-      }
+      val tokens = v.split("""[\._\-\+]""").toList
+      (tokens flatMap parseItem) ++ buildMeataData
+    }
 
-      def helper(s: Stream[Char]): Stream[(Separator, Item)] = {
-        if (s.isEmpty) Stream()
-        else {
-          val (sep, rem0) = parseSeparator(s)
-          sep match {
-            case Plus =>
-              Stream((sep, BuildMetadata(rem0.mkString)))
-            case _ =>
-              val (item, rem) = parseItem(rem0)
-              (sep, item) #:: helper(rem)
-          }
-        }
+    def parseItem(s: String): Option[Item] = {
+      if (s.isEmpty) None
+      else if (s.head.isDigit) {
+        val xs = s.toList
+        if (xs.forall(_.isDigit)) {
+          if (xs.size >= 10) Some(BigNumber(BigInt(s)))
+          else Some(Number(s.toInt))
+        } else Some(NonNum(s.toLowerCase)) // TODO use Scala.JS Locale library
+      } else {
+        val lc = s.toLowerCase // TODO use Scala.JS Locale library
+        if (s.head == '[' || s.head == ']') None
+        else if (lc == "max") Some(Max)
+        else if (lc == "min") Some(Min)
+        else Some(NonNum(lc))
       }
-
-      val (first, rem) = parseItem(s.toStream)
-      (first, helper(rem))
     }
   }
 
-  def postProcess(item: Item, tokens0: Stream[(Tokenizer.Separator, Item)]): Stream[Item] = {
-
-    val tokens =
-      // drop some '.0' under some conditions ???
-      if (isNumeric(item)) {
-        val nextNonDotZero = tokens0.dropWhile{case (Tokenizer.Dot, n: Numeric) => n.isEmpty; case _ => false }
-        if (nextNonDotZero.headOption.forall { case (sep, t) => sep != Tokenizer.Plus && !isMinMax(t) && !isNumeric(t) })
-          nextNonDotZero
-        else
-          tokens0
-      } else
-        tokens0
-
-    def ifFollowedByNumberElse(ifFollowedByNumber: Item, default: Item) = {
-      val followedByNumber = tokens.headOption
-        .exists{ case (Tokenizer.None, num: Numeric) if !num.isEmpty => true; case _ => false }
-
-      if (followedByNumber) ifFollowedByNumber
-      else default
+  def listCompare(first: List[Item], second: List[Item]): Int = {
+    def padNum(xs: List[Item], original: Int, next: Int): List[Item] = {
+      val (before, after) = xs.splitAt(original)
+      before ++ List.fill(next - original)(empty) ++ after
     }
-
-    val nextItem = item match {
-      case Literal("min") => Min
-      case Literal("max") => Max
-      case Literal("a") => ifFollowedByNumberElse(alphaQualifier, item)
-      case Literal("b") => ifFollowedByNumberElse(betaQualifier, item)
-      case Literal("m") => ifFollowedByNumberElse(milestoneQualifier, item)
-      case _ => item
+    val num1 = first.takeWhile(_.isNumber)
+    val num2 = second.takeWhile(_.isNumber)
+    (num1.size, num2.size) match {
+      case (x, y) if x == y =>
+        listCompare0(first, second)
+      case (x, y) if x > y  =>
+        listCompare0(first, padNum(second, y, x))
```

Comment:
Instead of dropping `.0` during parsing, I am padding lhs or rhs to have the same digits before comparison. I think this results to a more accurate comparison.